File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn merge_two_lists(
        mut list1: Option<Box<ListNode>>, 
        mut list2: Option<Box<ListNode>>
    ) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy;

        while list1.is_some() && list2.is_some() {
            let mut node1 = list1.take().unwrap();
            let mut node2 = list2.take().unwrap();

            if node1.val <= node2.val {
                list1 = node1.next.take();
                tail.next = Some(node1);
                list2 = Some(node2);
            } else {
                list2 = node2.next.take();
                tail.next = Some(node2);
                list1 = Some(node1);
            }

            tail = tail.next.as_mut().unwrap();
        }

        tail.next = if list1.is_some() 
        { list1 } else { list2 };

        dummy.next
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverse(self, head: Optional[ListNode]):
        curr = head
        prev = None
        while curr:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        return prev

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        curr = head
        count = k
        dummy = ListNode(0)
        tail = dummy
        group = curr

        while curr:
            if count == 1:
                next = curr.next
                curr.next = None
                rev = self.reverse(group)

                tail.next = rev
                for _ in range(k):
                    tail = tail.next
                group = next

                curr = next
                count = k
            else:
                curr = curr.next
                count -= 1

        tail.next = group
        
        return dummy.next